openfl/component/assigner/assigner.py

```
@@ -93,3 +102,47 @@ def get_aggregation_type_for_task(self, task_name):
         if "aggregation_type" not in self.tasks[task_name]:
             return None
         return self.tasks[task_name]["aggregation_type"]
+
+    @classmethod
+    def mode_based_filtering(cls, func):
+        """Decorator to filter task groups based on mode.
+
+        This decorator should be applied to define_task_assignments() method
+        in Assigner subclasses to handle mode-based filtering.
+        """
+
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            # First check if mode-based filtering is applicable
+            if hasattr(self, "mode"):
+                # Verify task_groups exists before attempting filtering
+                if not hasattr(self, "task_groups"):
+                    logger.warning(
+                        "Mode-based filtering was requested but no task_groups found. "
+                        "Skipping filtering. This might be intentional for custom assigners."
+                    )
+                    return func(self, *args, **kwargs)
+
+                if not self.task_groups:
+                    logger.warning(
```

@teoparvanov @kta-intel int this edge case should we err or warn 
line 138 errs post filtering if task groups are empty - these two behaviors are contradictory?

---

I suggest to "fail fast" in such cases, and raise an error, as on line 138. You could do so via an assertion, instead of a `ValueError` - see my comment below.

---

Done